### ISSUE TITLE
Upgrade our ZLib version for Windows

### ISSUE THREAD

**Issue by ianhays** [dotnet/corefx#25844]
We've missed a few releases of zlib and should update [our version of zlib](https://github.com/dotnet/corefx/tree/master/src/Native/Windows/clrcompression) from [1.2.8 to 1.2.11](https://www.zlib.net/ChangeLog.txt).

If possible, we should retain the perf improvements contributed by the Intel folks (cc: @bjjones).

@bjjones @ViktorHofer

**Comment by ianhays** [comment]
@joshfree @AtsushiKan

**Comment by joshfree** [comment]
@AtsushiKan I'd like to see if we can get this in the 2.1 wave.  Could you take a look?  Hopefully this is less than a day of work.

**Comment by AtsushiKan** [comment]
Made an attempt to do this - however, this causes eight test failures in System.IO.Compression.Tests ("ZLibException : The stream state of the underlying compression routine is inconsistent." thrown in a small set of tests that seem to be testing flushing, async and other strange scenarios.) This will require a deeper investigation to figure out what's going on. (See attached PR: https://github.com/dotnet/corefx/pull/28014)

**Comment by ianhays** [comment]
I've determined that our failures in updating are caused by https://github.com/madler/zlib/commit/b516b4bdd7c0c9f0858adfebf732089014f7b282

Essentially the error is that the 1.2.8 zlib code used to be very lax in verifying stream state between function calls, but with the above changeset state checks were added to a bunch of different call sites.

Searching for issues with the functions introduced in that changeset leads me to https://github.com/madler/zlib/issues/206.

I'm not sure yet, but the issue might be with how we use a zStream instantiated in managed memory and then wrap it in a `fixed` before passing the pointer to zlib. I think we should instead be instantiating the zstream in unmanaged memory so that it doesn't get moved by the GC which is breaking the pointer check added to zlib post-1.2.8
